Use mul! in Diagonal*Matrix

jishnub opened the pull request:
This provides a performance boost for matrices of numbers. See [the discourse post](https://discourse.julialang.org/t/possible-speedup-in-matrix-diagonal-products/68474)

On master
```julia
julia> A = ones(3000, 2000); DR = Diagonal(ones(size(A, 2))); DL = Diagonal(ones(size(A,1)));

julia> @btime $A * $DR;
  41.844 ms (4 allocations: 45.78 MiB)

julia> @btime $DL * $A;
  42.521 ms (2 allocations: 45.78 MiB)
```

After this PR

```julia
julia> @btime $A * $DR;
  28.670 ms (4 allocations: 45.78 MiB)

julia> @btime $DL * $A;
  29.377 ms (2 allocations: 45.78 MiB)
```

jishnub committed:
use mul! in Diagonal*Matrix of numbers

jishnub committed:
Add comment

MasonProtter commented:
Interesting finding! Is there any reason to not implement this with broadcasting though? Here are the numbers I'm seeing
```julia
julia> using BenchmarkTools: @btime

julia> using LinearAlgebra: Diagonal, mul!, similar, promote_op, transpose

julia> foreach((10, 50, 100, 500, 1_000, 5_000, 10_000)) do N
           M = N + rand(-5:5)
           A = rand(N, M)
           Dr = Diagonal(float.(1:M))

           println((;N, M))
           println("Current algorithm")
           @btime $A * $Dr
           println("Proposed algorithm")
           @btime $A *̂ $Dr
           println("Broadcasting")
           @btime $A .* transpose($Dr.diag)
           println()
       end
(N = 10, M = 14)
Current algorithm
  212.190 ns (3 allocations: 1.31 KiB)
Proposed algorithm
  192.421 ns (3 allocations: 1.31 KiB)
Broadcasting
  141.649 ns (1 allocation: 1.22 KiB)

(N = 50, M = 48)
Current algorithm
  2.612 μs (4 allocations: 18.92 KiB)
Proposed algorithm
  1.439 μs (4 allocations: 18.92 KiB)
Broadcasting
  1.346 μs (2 allocations: 18.83 KiB)

(N = 100, M = 99)
Current algorithm
  8.623 μs (4 allocations: 77.55 KiB)
Proposed algorithm
  4.386 μs (4 allocations: 77.55 KiB)
Broadcasting
  4.286 μs (2 allocations: 77.45 KiB)

(N = 500, M = 495)
Current algorithm
  212.060 μs (4 allocations: 1.89 MiB)
Proposed algorithm
  124.020 μs (4 allocations: 1.89 MiB)
Broadcasting
  126.190 μs (2 allocations: 1.89 MiB)

(N = 1000, M = 999)
Current algorithm
  1.336 ms (4 allocations: 7.62 MiB)
Proposed algorithm
  799.228 μs (4 allocations: 7.62 MiB)
Broadcasting
  801.688 μs (2 allocations: 7.62 MiB)

(N = 5000, M = 5001)
Current algorithm
  64.819 ms (4 allocations: 190.77 MiB)
Proposed algorithm
  43.458 ms (4 allocations: 190.77 MiB)
Broadcasting
  43.699 ms (2 allocations: 190.77 MiB)

(N = 10000, M = 9997)
Current algorithm
  266.747 ms (4 allocations: 762.71 MiB)
Proposed algorithm
  160.720 ms (4 allocations: 762.71 MiB)
Broadcasting
  159.481 ms (2 allocations: 762.71 MiB)
```
It seems that just using broadcast is simpler and just as fast or faster.


_____

```julia
julia> versioninfo()
Julia Version 1.6.1
Commit 6aaedecc44 (2021-04-23 05:59 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: AMD Ryzen Threadripper 2990WX 32-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, znver1)
Environment:
  JULIA_NUM_THREADS = 26
```

MasonProtter commented:
Normally, matrix multiplication involves a mixture of broadcast **and** reduction, but involving a `Diagonal` explicitly removes the reduction part and you're left with just an elementwise operation.

N5N3 commented:
In fact `mul!()`/`rmul!()`/`lmul!()` for Diagonal is implemented via broadcast.

There should be no runtime performance difference between 3-arg `mul!()` and manual broadcast, as the `alpha` and `beta` are propagated as const.

MasonProtter commented:
Hm. Then why is there such a big performance difference here?

N5N3 commented:
The difference holds only for small martix. It might comes from (I didn't verify):
1. You manual broadcast use lazy `transpose`, while `mul!` use `premutedims`, which is create a new `Matrix`.
2. `mul!(z,x,y::Diagnal)` calculate z .= x .* permutedims(y.diag) .+ false. Maybe the `.+` was not omited by the optimizer?

MasonProtter commented:
Hm. That eager `permutedims` Should probably be revisited.

jishnub committed:
runtime checks to avoid ambiguities

N5N3 commented:
`@code_typed` shows that their loop body have almost the same IR.
except `mul!` with some `dead code`:
```julia
│     %308 = Base.mul_float(%304, %307)::Float64
│     %309 = Base.copysign_float(0.0, %308)::Float64
│     %310 = Base.ifelse(true, %308, %309)::Float64
│     %311 = Base.add_float(1.0, %310)::Float64
│     %312 = Base.ifelse(false, %311, %310)::Float64 
```
LLVM should be able to omit them.
Besides that `mul!` seems to perform more checks before the loop kernal, maybe this also contributes to the speed gap for small matrix.

jishnub commented:
The difference for small matrices seems to be entirely due to `permutedims` being an array reshape that allocates:

```julia
julia> A = ones(10, 14); DR = Diagonal(ones(size(A, 2)));

julia> @btime permutedims($DR.diag);
  98.516 ns (2 allocations: 96 bytes)

julia> @btime transpose($DR.diag);
  2.674 ns (0 allocations: 0 bytes)

julia> @btime $A * $DR;
  439.354 ns (3 allocations: 1.31 KiB)

julia> @btime $A .* transpose($DR.diag);
  344.735 ns (1 allocation: 1.22 KiB)
```

jishnub committed:
Force destination to be a Matrix

N5N3 commented:
It seems broadcast based `mul!()` was limited to `StridedMatix`. For example:
```julia
julia> @which mul!(zeros(100,100), Diagonal(1:100), Symmetric(randn(100,100)), true, false)
mul!(C::AbstractMatrix, A::AbstractVecOrMat, B::AbstractVecOrMat, alpha::Number, beta::Number) in LinearAlgebra at C:\Users\MYJ\AppData\Local\Programs\Julia-1.7.0-beta4\share\julia\stdlib\v1.7\LinearAlgebra\src\matmul.jl:301
julia> @which mul!(zeros(100,100), Diagonal(1:100), randn(100,100), true, false)
mul!(out::AbstractMatrix, A::Diagonal, in::StridedMatrix{T} where T, alpha::Number, beta::Number) in LinearAlgebra at C:\Users\MYJ\AppData\Local\Programs\Julia-1.7.0-beta4\share\julia\stdlib\v1.7\LinearAlgebra\src\diagonal.jl:370
```
In the above case maybe call optimized `copyto!` first, then call `lmul!` is a better choice?

jishnub commented:
Would it make sense to improve `mul!` in a separate PR? Changing `StrideMatrix` to `AbstractMatrix` might need to be tested thoroughly.

N5N3 commented:
Both seems OK to me.
At present, in test.jl, `lmul!`, `rmul!` and `mul!` appears  11, 10, and 13 times, seperately.
While `*` appears 170 times.
Even if we assume that half of them are the reference for comparison, `*` is tested much more thoroughly.
Without this PR, we can say that `lmul!` and `rmul!` are well tested via `*`.
After this PR, things turns to be `mul!` is tested via `*`, while `lmul!` and `rmul!` not.
So it might be ok to do this change in this PR?

N5N3 commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -219,10 +219,32 @@ end
 (*)(D::Diagonal, B::AbstractTriangular) =
```
broadcast based `mul!` also works for `AbstractTriangular`.
optimize this two in this PR as well?

dkarrasch commented:
> Hm. That eager `permutedims` Should probably be revisited. Seems like a waste of allocations that was done in the past because we didn’t handle `struct`s containing mutable data well.

No, we use `permutedims` because `transpose` operates recursively, but all we want is to "reshape" the vector into a row vector for broadcasting. If there is another (lazy) way of telling broadcasting that the vector should be interpreted as 1xn, then we can go with that. Otherwise, products with matrices of matrices will just go wrong.

jishnub committed:
Require one-based indices

jishnub commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -219,10 +219,32 @@ end
+function (*)(A::AbstractMatrix, D::Diagonal)
+    Tdest = promote_op(*, eltype(A), eltype(D.diag))
+    if eltype(A) <: Number && eltype(D) <: Number
+        nA, nD = size(A, 2), length(D.diag)
```
Added

MasonProtter commented:
> No, we use permutedims because transpose operates recursively, but all we want is to "reshape" the vector into a row vector for broadcasting.

That shouldn’t be a problem if the method is restricted to operating on `Diagonal{T, Vector{T}} where T <: Number` though right? 

I guess if someone made some wacky number type that could be `transpose`’d that could be a problem?

N5N3 commented:
Since this PR accelerate `(*)(A::AbstractMatrix, D::Diagonal)`, I think at least an optimized `(*)(A::AbstractTriangular, D::Diagonal)` could be implemented via `(*)(A::AbstractMatrix, D::Diagonal)` to avoid data copy. 
For example:
```julia
@inline function _copy_diag!(data, diag)
    for i in 1:length(diag)
        data[i,i] = D.diag[i]
    end
    data
end
(*)(A::UpperTriangular, D::Diagonal) = UpperTriangular(A.data * D)
(*)(A::UnitUpperTriangular, D::Diagonal) = UpperTriangular(_copy_diag!(A.data * D, D.diag))
```

N5N3 commented:
BTW, I found that the above pattern also works for `(l/r)mul!`. 
We can put all these definitions together to simplify code:
```julia
for Tri in (:UpperTriangular, :LowerTriangular)
    UTri = Symbol(:Unit, Tri)
    for fun in (:*, :rmul!)
        @eval ($fun)(A::$Tri, D::Diagonal) = $Tri($fun(A.data, D))
        @eval ($fun)(A::$UTri, D::Diagonal) = $Tri(_copy_diag!($fun(A.data, D), D.diag))
    end
    for fun in (:*, :lmul!)
        @eval ($fun)(D::Diagonal, A::$Tri) = $Tri($fun(D, A.data))
        @eval ($fun)(D::Diagonal, A::$UTri) = $Tri(_copy_diag!($fun(D, A.data), D.diag))
    end
    @eval mul!(out::$Tri, D::Diagonal, A::$Tri) = $Tri(mul!(out.data, D, A.Data))
    @eval mul!(out::$Tri, D::Diagonal, A::$UTri) = $Tri(_copy_diag!(mul!(out.data, D, A.Data), D.diag))
    @eval mul!(out::$Tri, A::$Tri, D::Diagonal) = $Tri(mul!(out.data, A.Data, D))
    @eval mul!(out::$Tri, A::$UTri, D::Diagonal) = $Tri(_copy_diag!(mul!(out.data, A.Data, D), D.diag))
end
```
The above code support seperate optimization for `*()` and `(r/l)mul!`.
And it's more easy to cover all cases, as at present only `rmul!` is optimized for `Union{LowerTriangular,UpperTriangular}`
https://github.com/JuliaLang/julia/blob/7ce858c2ddcc889d682603d85a544af6169f5516/stdlib/LinearAlgebra/src/diagonal.jl#L271
`lmul!` for that has been missed for years.

@dkarrasch Should this be included in a seperate PR?

jishnub commented:
Perhaps this should be specialized to cases where `A.data * D` is fast? This shouldn't hit slow `AbstractArray` multiplication methods. The present methods that rely on a `copy` get around these methods for the cases where `A.data` is not a `Matrix`. Even the method added in this PR might need to be restricted (perhaps to `StridedMatrix`?).

Currently on master
```julia
julia> A = LowerTriangular(reshape(1:1600, 40, 40)); D = Diagonal(ones(size(A,2)));

julia> @btime $A * $D;
  8.931 μs (3 allocations: 12.72 KiB)

julia> @btime $A.data * $D;
  2.308 μs (3 allocations: 12.72 KiB)

julia> @btime mul!(similar($A.data, size($A.data)), $A.data, $D);
  61.452 μs (7 allocations: 12.95 KiB)
```

With this PR, we're hitting `mul!` in  `A.data * D`, which will be slow for `AbstractArray`s in general.

N5N3 commented:
I believe that `rmul!` for `AbstractTriangular` was implemented in that way because It allows better vectorization and memory access at the sacrifice of doubled multiplication. 
(In most cases it accelerate the code, except for a arraytype with extremely slow getindex and setindex!, for that case I think collect it first is always a good choise?)

Anyway, I think `r/lmul!`, 3-arg `mul!` and `*` should not hit genetic matmul.
Although broadcast might be sub-optimal in some cases, it should be much faster than genetic matmul.

jishnub commented:
Currently 

```julia
julia> A = reshape(1:1600, 40, 40); B = Diagonal(ones(Int, size(A,2))); C = similar(A, eltype(A), size(A));

julia> @which mul!(C, A, B, true, false)
mul!(C::AbstractMatrix{T} where T, A::AbstractVecOrMat{T} where T, B::AbstractVecOrMat{T} where T, alpha::Number, beta::Number) in LinearAlgebra at /home/jb6888/julia/julia-1.6.2/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:302

julia> @which mul!(C, collect(A), B, true, false)
mul!(out::AbstractMatrix{T} where T, in::StridedMatrix{T} where T, A::Diagonal, alpha::Number, beta::Number) in LinearAlgebra at /home/jb6888/julia/julia-1.6.2/share/julia/stdlib/v1.6/LinearAlgebra/src/diagonal.jl:356
```

The first method hits the generic matmul. Perhaps this may be implemented through broadcasting?

N5N3 commented:
If I have not made a mistake:
The difference between `mul!` and `r/lmul!` based `*(A::AbstractMatrix,D::Diagnal)` is:
1. `r/lmul!`  --> copyto!(similar(A, Tdest, size(A)), A) .*=  D.diag
2. `mul!`  --> similar(A, Tdest, size(A)) .=  A .* D.diag
In most cases, type promotion should be vectorizable, since all elements in A is accessed only once, should pre-promotion offer any performance gain?

BTW, for this example
```julia
julia> A = LowerTriangular(reshape(1:1600, 40, 40)); D = Diagonal(ones(size(A,2)));
julia> @btime $A.data * $D;
  2.331 μs (3 allocations: 12.72 KiB)
julia> @btime mul!(similar($A.data, size($A.data)), $A.data, $D);
  3.239 μs (3 allocations: 12.72 KiB)
```
Their return type are not the same: `A.data * D` returns `Matrix{Float64}`, while `mul!` version returns  `Matrix{Int64}`. So the `regression` comes from extra trunction.
On my PC:
```julia
julia> @btime $A.data * $D;
  1.950 μs (3 allocations: 12.72 KiB)
julia> @btime mul!(similar($A.data, size($A.data)), $A.data, $D);
  2.422 μs (3 allocations: 12.72 KiB)
julia> @btime mul!(similar($A.data, Float64, size($A.data)), $A.data, $D);
  1.250 μs (3 allocations: 12.72 KiB)
```

dkarrasch commented:
I don't follow why we need this strange `Number`s distinction. I think the main difference between `mul!` and `[r/l]mul!`-based implementations is that in the first case we only allocate a similar array and write the result into it, whereas in the latter case we allocate a similar matrix, then fill it with values of `A` and then pass over it again to multiply each value. So, the only difference is whether we fill the destination array with the (promoted) values of `A` and then modify inplace or whether we write into an `undef` destination array. But why the distinction with `Number`s? Why not just the size checks, the allocation of the destination array and then `mul!` into it?

dkarrasch commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -219,10 +219,34 @@ end
+function (*)(A::AbstractMatrix, D::Diagonal)
+    Tdest = promote_op(*, eltype(A), eltype(D.diag))
+    if eltype(A) <: Number && eltype(D) <: Number
+        Base.require_one_based_indexing(A)
```
I don't see where one-based indexing matters here. There is no indexing going on, no `for i in 1:n` and alike.

N5N3 replied:
Diagonal’s diag vector is restricted to one-base indexing.

dkarrasch replied:
Yes, but `rmul!` probably has it because it involves indexing which starts at 1. I assume the check is performed inside `mul!` and `rmul!` anyways.

dkarrasch commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -219,10 +219,34 @@ end
 (*)(D::Diagonal, B::AbstractTriangular) =
     lmul!(D, copy_oftype(B, promote_op(*, eltype(B), eltype(D.diag))))
 
-(*)(A::AbstractMatrix, D::Diagonal) =
-    rmul!(copy_similar(A, promote_op(*, eltype(A), eltype(D.diag))), D)
-(*)(D::Diagonal, A::AbstractMatrix) =
-    lmul!(D, copy_similar(A, promote_op(*, eltype(A), eltype(D.diag))))
```
It seems like the whole performance improvement can be achieved by replacing these two methods with
```julia
(*)(A::AbstractMatrix, D::Diagonal) =
    mul!(similar(A, promote_op(*, eltype(A), eltype(D.diag)), size(A)), A, D)
(*)(D::Diagonal, A::AbstractMatrix) =
    mul!(similar(A, promote_op(*, eltype(A), eltype(D.diag)), size(A)), D, A)
```
In fact, the second isn't even necessary and could be removed(!), since the fallbacks handle everything.

dkarrasch commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -219,10 +219,34 @@ end
-    rmul!(copy_similar(A, promote_op(*, eltype(A), eltype(D.diag))), D)
-(*)(D::Diagonal, A::AbstractMatrix) =
-    lmul!(D, copy_similar(A, promote_op(*, eltype(A), eltype(D.diag))))
+# Using mul! instead of (l/r)mul! provides a performance boost by avoiding a copy.
+# This is not possible in general, but works if the elements are numbers
+# See #42321
+# Using runtime eltype checks instead of dispatch to avoid ambiguities
+function (*)(A::AbstractMatrix, D::Diagonal)
+    Tdest = promote_op(*, eltype(A), eltype(D.diag))
+    if eltype(A) <: Number && eltype(D) <: Number
+        Base.require_one_based_indexing(A)
+        nA, nD = size(A, 2), length(D.diag)
+        if nA != nD
+            throw(DimensionMismatch("second dimension of A, $nA, does not match the first of D, $nD"))
+        end
+        return mul!(similar(A, Tdest, size(A)), A, D)
+    end
+    return rmul!(copy_similar(A, Tdest), D)
+end
+function (*)(D::Diagonal, A::AbstractMatrix)
+    Tdest = promote_op(*, eltype(A), eltype(D.diag))
+    if eltype(A) <: Number && eltype(D) <: Number
+        Base.require_one_based_indexing(A)
+        nA, nD = size(A, 1), length(D.diag)
+        if nA != nD
+            throw(DimensionMismatch("second dimension of D, $nD, does not match the first of B, $nA"))
+        end
+        return mul!(similar(A, Tdest, size(A)), D, A)
+    end
+    return lmul!(D, copy_similar(A, Tdest))
+end
```
It seems like the whole performance improvement can be achieved by replacing the above two methods by
```suggestion
(*)(A::AbstractMatrix, D::Diagonal) =
    mul!(similar(A, promote_op(*, eltype(A), eltype(D.diag)), size(A)), A, D)
(*)(D::Diagonal, A::AbstractMatrix) =
    mul!(similar(A, promote_op(*, eltype(A), eltype(D.diag)), size(A)), D, A)
```
In fact, the second isn't even necessary and could be removed(!), since the fallbacks handle everything. This avoids the second pass over the destination array.

N5N3 replied:
`+(::AbstractArray, ::Bool)` throw MethodError now.
So current `mul!` should be limited to Number Eltype.
We can define something like `*ₛ` to partially handle it:
```julia
function +ₛ end
Base.broadcasted(::typeof(+ₛ), ::Bool, bc) = bc
Base.broadcasted(::typeof(+ₛ), bc, ::Bool) = bc
Base.broadcasted(::typeof(+ₛ), bc₁, bc₂) = Base.broadcasted(+, bc₁, bc₂)
# what should we set in out for non-number eltype if alpha == beta == 0
Base.broadcasted(::typeof(+ₛ), ::Bool, ::Bool) = false 
```

dkarrasch replied:
Where do you see that failure?

dkarrasch replied:
I'm also not seeing extra allocations:
```julia
julia> @btime mul!($(similar(A)), $A, $DR);
  5.636 ms (2 allocations: 96 bytes)
```
We may have a few extra operations, but no extra allocations because `mul!` broadcasts right into the destination. Why can't we handle the alpha=beta=0 case? I assume this is all well tested, isn't it?

N5N3 replied:
```julia
julia> A = Diagonal([[1 2;3 4],[1 2;3 4]])
2×2 Diagonal{Matrix{Int64}, Vector{Matrix{Int64}}}:
 [1 2; 3 4]      ⋅
     ⋅       [1 2; 3 4]

julia> B = [1,1]
2-element Vector{Int64}:
 1
 1

julia> A * B
2-element Vector{Matrix{Int64}}:
 [1 2; 3 4]
 [1 2; 3 4]

julia> C =  A * B
2-element Vector{Matrix{Int64}}:
 [1 2; 3 4]
 [1 2; 3 4]

julia> mul!(C,A,B)
ERROR: MethodError: no method matching +(::Matrix{Int64}, ::Bool)
```

dkarrasch replied:
Ok, then I suggest we simply add a few 3-arg `mul!` methods:
```julia
@inline mul!(out::AbstractVector, A::Diagonal, in::AbstractVector) =
    out .= A.diag .* in
@inline mul!(out::AbstractMatrix, A::Diagonal, in::AbstractMatrix) =
    out .= A.diag .* in
@inline mul!(out::AbstractMatrix, A::Diagonal, in::Adjoint{<:Any,<:AbstractVecOrMat}) =
    out .= A.diag .* in
@inline mul!(out::AbstractMatrix, A::Diagonal, in::Transpose{<:Any,<:AbstractVecOrMat}) =
    out .= A.diag .* in

@inline mul!(out::AbstractMatrix, in::AbstractMatrix, A::Diagonal) =
    out .= in .* permutedims(A.diag)
@inline mul!(out::AbstractMatrix, in::Adjoint{<:Any,<:AbstractVecOrMat}, A::Diagonal) =
    out .= in .* permutedims(A.diag)
@inline mul!(out::AbstractMatrix, in::Transpose{<:Any,<:AbstractVecOrMat}, A::Diagonal) =
    out .= in .* permutedims(A.diag)
```

jishnub replied:
This seems to be a good approach, and I'll admit the restriction to numbers didn't have any fundamental basis other than avoiding missing methods. I'll try adding the suggested methods in over the weekend and see if things work

jishnub replied:
I have added the three element `mul!`, but this doesn't solve the issue with the five-element `mul!` that still remains.

```julia
julia> A = reshape([[1 2; 3 4], zeros(Int,2,2), zeros(Int, 2, 2), [5 6; 7 8]], 2, 2)
2×2 Matrix{Matrix{Int64}}:
 [1 2; 3 4]  [0 0; 0 0]
 [0 0; 0 0]  [5 6; 7 8]

julia> D = Diagonal(A)
2×2 Diagonal{Matrix{Int64}, Vector{Matrix{Int64}}}:
 [1 2; 3 4]      ⋅     
     ⋅       [5 6; 7 8]

julia> mul!(similar(A), A, D, true, false)
ERROR: MethodError: no method matching +(::Matrix{Int64}, ::Bool)

julia> mul!(similar(A), A, D)
2×2 Matrix{Matrix{Int64}}:
 [7 10; 15 22]  [0 0; 0 0]
 [0 0; 0 0]     [67 78; 91 106]
```

dkarrasch reviewed (changes requested):
I think I found a much simpler solution. BTW, do we really need more tests? I'm sure we have Array*Diagonal covered. Maybe not the non-strided case though?

jishnub committed:
don't specialize * for numbers

N5N3 commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -198,31 +198,71 @@ Base.literal_pow(::typeof(^), D::Diagonal, valp::Val) =
     return D.diag .* V
 end
 
+# Optimized multiplication between Triangular and Diagonal matrices
```
An extended version of this, which includes `/` related calls, is under going in #42343. So these 2 PRs seem to conflict with each other.

jishnub replied:
Ah I see, so maybe I'll remove these from this PR

jishnub replied:
I've removed these from this PR

jishnub committed:
don't specialize for triangular

jishnub committed:
don't use in as a variable

jishnub committed:
use branch in five-element mul

jishnub committed:
Add test for 5-element mul

jishnub committed:
simplify size mismatch error check

jishnub committed:
fix mul! broadcasting

jishnub commented:
@dkarrasch @N5N3 could you take a look now? I have incorporated the changes suggested. I have added some extra methods for size check and the actual multiplication to avoid replicating code, hope this is fine.

N5N3 reviewed (commented):
Just some recommendations:
1. There seems so reason to retain the `l\rmul!` in L245, L255. We can make them call 3-arg `mul!` (to reuse the input check).
2. It looks good to have an optimized `mul!` for `Diagonal` outputs. (L374 is suboptimal in this case). With it, we can omit L327 and L328.
3. ~Add `require_one_based_indexing` in the size check function?~
4. Edited: L339 seems suboptimal for `eltype <:  matrix`. As `mul!(A,D,B)` fallbacks to `A .= (D.diag .* B) .* true`, which allocates twice unnecessarily in this case.

N5N3 commented:
My fault, didn't notice that we have optimize `*(::Matrix,::Matrix,::Number)`
```julia
mat_mat_scalar(A::StridedMaybeAdjOrTransMat, B::StridedMaybeAdjOrTransMat, γ) =
    _mat_mat_scalar(A, B, γ)

function _mat_mat_scalar(A, B, γ)
    T = promote_type(eltype(A), eltype(B), typeof(γ))
    C = similar(A, T, axes(A,1), axes(B,2))
    mul!(C, A, B, γ, false)
end
```

dkarrasch commented:
> Just some recommendations:
> 
>     1. There seems so reason to retain the `l\rmul!` in L245, L255. We can make them call 3-arg `mul!` (to reuse the input check).

I agree with this one.

>     2. It looks good to have an optimized `mul!` for `Diagonal` outputs. (L374 is suboptimal in this case). With it, we can omit L327 and L328.

I agree with this one, as well.

>     3. Add `require_one_based_indexing` in the size check function?

This is unnecessary. One-based index checking is only necessary when we index into an array in a loop that has a loop variable `i in 1:n`. Otherwise, there is no need for such a check.

>     4. L339 seems suboptimal for `eltype <:  matrix`. As 3-arg `mul!(A,D,B)` calls `A .= D.diag .* B . * 1`, which allocates twice unnecessarily in this case.

Up to the obsolete multiplication with 1 there is no issue with memory allocation. I don't think we want to introduce more runtime checks here.

dkarrasch commented:
> My fault, didn't notice that we have optimize `*(::Matrix,::Matrix,::Number)` via `mat_mat_scalar`

It has nothing to do with that method. It is using broadcasting, which internally writes out the loops and writes into the destination array within that loop. So that operation is in-place, as you can check with BenchmarkTools.

N5N3 commented:
For Number element I aggree with you. But for Matrix element:
```julia
julia> a = [randn(3,3) for i in 1:10]; b = [randn(3,3) for i in 1:10];
julia> @btime $a .* $b;
  507.216 ns (11 allocations: 1.38 KiB)
julia> @btime $a .* $b .* 1;
  895.122 ns (22 allocations: 2.67 KiB)
julia> @btime @. $a * $b * 1;
  515.544 ns (11 allocations: 1.38 KiB)
```

dkarrasch commented:
> So the allocation does happen twice.

Are you sure? IIRC, that special multiplication recognizes constant `true`s and avoids performung multiplication by `true = 1`.

N5N3 commented:
```julia
julia> g(a, b) = a .* b .* true
g (generic function with 1 method)
julia> @btime g($a,$b);
  880.000 ns (21 allocations: 2.62 KiB)
julia> g2(a, b) = a .* b
g2 (generic function with 1 method)
julia> @btime g2($a,$b);
  509.896 ns (11 allocations: 1.38 KiB)
```
I think the above example is enough.
At present we replace `a * true` with `a` by llvm. It seems works only for `number * number`

dkarrasch commented:
> So also add something somthing like?
> 
> ```julia
> *(a::AbstractArray{<:Number},b::Number,c::Number) = b * c * a
> *(a::Number,b::AbstractArray{<:Number},c::Number) = a * c * b
> ```

Better not to change the order of multiplication and use broadcast:
```julia
*(a::AbstractArray{<:Number},b::Number,c::Number) = a .* b .* c
*(a::Number,b::AbstractArray{<:Number},c::Number) = a .* b .* c
```
I'm not sure how sustainable such definitions are, though. These would help with the array of arrays case, but what about the array of arrays of arrays? Feels a bit weird to mess with such a basic operation.

jishnub committed:
mul! with Diagonal output and add zero and one

jishnub commented:
If we add an `isone` check while broadcasting over `*ₛ` 

```julia
Broadcast.broadcasted(::typeof(*ₛ), out, beta) =
    iszero(beta::Number) ? false :
    isone(beta::Number) ? broadcasted(identity, out) : broadcasted(*, out, beta)
```

 we may get rid of the extra allocation:

```julia
julia> D
2×2 Diagonal{Matrix{Int64}, Vector{Matrix{Int64}}}:
 [7 15; 10 22]      ⋅     
     ⋅          [67 91; 78 106]

julia> A = collect(D)
2×2 Matrix{Matrix{Int64}}:
 [7 15; 10 22]  [0 0; 0 0]
 [0 0; 0 0]     [67 91; 78 106]

julia> B = similar(A);

julia> @btime mul!($B, $D, $A);
  337.647 ns (4 allocations: 384 bytes)

julia> @btime $B .= $D.diag .* $A;
  323.026 ns (4 allocations: 384 bytes)

julia> @btime mul!($B, $D, $A, true, false);
  336.775 ns (4 allocations: 384 bytes)
```

jishnub commented:
I've also specialized `zero` and `one` for `Diagonals`, so now these functions work for nested diagonals:

```julia
julia> D = Diagonal([collect(reshape(1:4, 2, 2)), collect(reshape(5:8, 2, 2))])
2×2 Diagonal{Matrix{Int64}, Vector{Matrix{Int64}}}:
 [1 3; 2 4]      ⋅     
     ⋅       [5 7; 6 8]

julia> D2 = Diagonal([D, D]);

julia> D2 + zero(D2) == D2
true

julia> D2 * one(D2) == D2
true
```

N5N3 commented:
I think the behaviour change of `.*ₛ` is acceptable. 
In most case it is used to dispatch `alpha` and `beta`, so I think the performance influence won't be large?

BTW, if someone dislike this modification. 
Maybe we can use `MulAddMul` to achieve the same goal:
```julia
@noinline ___muldiag!(out,A,B,f::MulAddMul{<:Any,true}) = @. out = f(A * B) # out might be undef
@noinline ___muldiag!(out,A,B,f::MulAddMul) = @. out = f(A * B,out) 
@inline __muldiag!(out, D::Diagonal, B, f) = ___muldiag!(out, D.diag, B, f)
@inline __muldiag!(out, A, D::Diagonal, f) = ___muldiag!(out, A, permutedims(D.diag), f)
@inline __muldiag!(out::Diagonal, D1::Diagonal, D2::Diagonal, f) = Diagonal(___muldiag!(out.diag, D1.diag, D2.diag, f))
@inline function _muldiag!(out, A, B, alpha, beta)
    _muldiag_size_check(out, A, B)
    iszero(alpha) && return _rmul_or_fill!(out, beta)
    __muldiag!(out, A, B, MulAddMul(alpha, beta))
end
```
On my desktop it tooks about 20ns to do this dynamic dispatch.
So the performance does regress slightly compared with this PR in some cases, such as non-const `beta == 0`.

jishnub committed:
bugfix in mul for diagonal output

jishnub committed:
Specialize oneunit for Diagonal

jishnub committed:
replace conflicting definition for one(::Diagonal)

jishnub commented:
@dkarrasch I've made most of the changes suggested, could you take a look at this?

staticfloat commented:
@dkarrasch @jishnub I am observing a pretty substantial slowdown on the LinearAlgebra tests due to this commit.

Checking out `a8db7510b97571cb9ad6cc89fc2fae527c741ea6` and the commit before it, I see (for instance) the `LinearAlgebra/addmul` tests go from 280s to 340s.  This is when running on a `Intel(R) Xeon(R) CPU E3-1241 v3 @ 3.50GHz`.

N5N3 reviewed (commented):
Just a quick glance.

dkarrasch commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -220,29 +238,12 @@ end
-    B .= D.diag .* B
-    return B
-end
+rmul!(A::AbstractMatrix, D::Diagonal) = mul!(A, A, D)
+lmul!(D::Diagonal, B::AbstractVecOrMat) = mul!(B, D, B)
```
Thanks for taking a look, @N5N3! I vaguely remember that you proposed in a similar situation the fix by inlining the rhs. Why doesn't that help here?

N5N3 replied:
IIUC, our broadcast unaliases the `src` from the `dest` to avoid memory contention, which prevents LLVM to prove the `dest === src`.
Our for-loop based implement doesn't include such unaliasing stage, thus `@inline` is enough to enable simd.
